suite/tests/client-interface/emulation_api_simple.dll.c

```diff
@@ -84,77 +331,95 @@ event_instruction_change(void *drcontext, void *tag, instrlist_t *bb, bool for_t
-
-        dr_save_reg(drcontext, bb, instr, DR_REG_X26, SPILL_SLOT_1);
-        dr_save_reg(drcontext, bb, instr, DR_REG_X27, SPILL_SLOT_2);
-
-        opnd_t scratch0 = opnd_create_reg(DR_REG_X26);
-        opnd_t scratch1 = opnd_create_reg(DR_REG_X27);
+        if (drmgr_is_emulation_end(instr)) {
+            in_emulation = false;
+        }
+    }
+    return DR_EMIT_DEFAULT;
+}
 
-        /* scratch0 = !src0
-         * XXX i#2440 AArch64 missing INSTR_CREATE macros (INSTR_CREATE_orn)
-         */
-        instr_t *not0 = instr_create_1dst_4src(drcontext, OP_orn, scratch0,
-                                               opnd_create_reg(DR_REG_XZR), src0,
-                                               OPND_CREATE_LSL(), OPND_CREATE_INT(0));
-        instrlist_preinsert(bb, instr, instr_set_translation(not0, raw_instr_pc));
+static void
+record_instr_fetch_orig(instr_t *instr)
+{
+    DR_ASSERT(should_fully_emulate_instr(instr) || should_partly_emulate_instr(instr));
+    DR_ASSERT((ptr_uint_t)instr_get_note(instr) != derived_marker);
+}
 
-        /* scratch1 = !src1
-         * XXX i#2440 AArch64 missing INSTR_CREATE macros (INSTR_CREATE_orn)
-         */
-        instr_t *not1 = instr_create_1dst_4src(drcontext, OP_orn, scratch1,
-                                               opnd_create_reg(DR_REG_XZR), src1,
-                                               OPND_CREATE_LSL(), OPND_CREATE_INT(0));
-        instrlist_preinsert(bb, instr, instr_set_translation(not1, raw_instr_pc));
+static void
+record_instr_fetch_unchanged(instr_t *instr)
+{
+    DR_ASSERT(!should_partly_emulate_instr(instr) && !should_fully_emulate_instr(instr));
+    DR_ASSERT((ptr_uint_t)instr_get_note(instr) != derived_marker);
+}
 
-        /* scratch0 = scratch0 | scratch1
-         * XXX i#2440 AArch64 missing INSTR_CREATE macros (INSTR_CREATE_orr)
-         */
-        instr_t *or_i =
-            instr_create_1dst_4src(drcontext, OP_orr, scratch0, scratch0, scratch1,
-                                   OPND_CREATE_LSL(), OPND_CREATE_INT(0));
-        instrlist_preinsert(bb, instr, instr_set_translation(or_i, raw_instr_pc));
+static void
+record_data_addresses_orig(instr_t *instr)
+{
+    DR_ASSERT(should_fully_emulate_instr(instr));
+    DR_ASSERT(!should_partly_emulate_instr(instr));
+    DR_ASSERT((ptr_uint_t)instr_get_note(instr) != derived_marker);
+}
 
-        /* dst = !scratch0
-         * XXX i#2440 AArch64 missing INSTR_CREATE macros (INSTR_CREATE_orn)
-         */
-        instr_t *not2 =
-            instr_create_1dst_4src(drcontext, OP_orn, dst, opnd_create_reg(DR_REG_XZR),
-                                   scratch0, OPND_CREATE_LSL(), OPND_CREATE_INT(0));
-        instrlist_preinsert(bb, instr, instr_set_translation(not2, raw_instr_pc));
+static void
+record_data_addresses_derived(instr_t *instr)
+{
+    DR_ASSERT(!should_partly_emulate_instr(instr) && !should_fully_emulate_instr(instr));
```

Comment:
Since the instr is being emulated for the "partly" case, we should *not* see the original instr here but instead the replacement: and thus should_partly_emulate_instr must return false.  I added a comment.